File: C++/섹션08. A형 문제 풀이 (~2022 상반기)/15. 드래곤 커브 (코드트리, 2018 상반기 오후 1번).cpp
#include <stdio.h>

#define MAX (100 + 20)

int T;

int N;
int MAP[MAX][MAX];
int dragonPosition[2000]; // 2^10 보다 큰 값

// 오른쪽, 위쪽, 왼쪽, 아래쪽
int dr[] = { 0, -1, 0, 1 };
int dc[] = { 1, 0, -1, 0 };

void makeDragonCurve()
{
	scanf("%d", &N);

	for (int i = 0; i < N; i++)
	{	
		int r, c, d, g; // x = r, y = c;

		scanf("%d %d %d %d", &r, &c, &d, &g);

		dragonPosition[0] = d;

		int length = 1;
		for (int step = 1; step <= g; step++)
		{
			for (int i = length + 1; i <= length * 2; i++)
				dragonPosition[i - 1] = (dragonPosition[length * 2 - i] + 1) % 4;

			// for debug
			// for (int i = 0; i < length * 2; i++) printf("%d, ", dragonPosition[i]);
			// putchar('\n');

			length *= 2;
		}

		MAP[r][c] = 1;

		int moveCount = 1 << g; // move = 2^g
		for (int k = 0; k < moveCount; k++)
		{
			r = r + dr[dragonPosition[k]];
			c = c + dc[dragonPosition[k]];

			MAP[r][c] = 1;
		}
	}
}

int countSquare()
{
	int sum = 0;
	for (int r = 0; r < 100; r++)
		for (int c = 0; c < 100; c++)
			if (MAP[r][c] + MAP[r + 1][c] + MAP[r][c + 1] + MAP[r + 1][c + 1] == 4) sum++;

	return sum;
}

int main()
{
	// scanf("%d", &T);
	T = 1;
	for (int tc = 1; tc <= T; tc++)
	{
		makeDragonCurve();

		printf("%d\n", countSquare());
	}

	return 0;
}

File: Python/섹션08. A형 문제 풀이 (~2022 상반기)/15. 드래곤 커브 (코드트리, 2018 상반기 오후 1번).py
MAX = 100 + 20

MAP = [[0] * MAX for _ in range(MAX)]
dragonPosition = [0] * 2000  # 2^10보다 큰 값 

# 오른쪽, 위쪽, 왼쪽, 아래쪽
dr = [0, -1, 0, 1]
dc = [1, 0, -1, 0]

def make_dragon_curve():    
    N = int(input())

    for _ in range(N):
        r, c, d, g = map(int, input().split())

        dragonPosition[0] = d
        
        length = 1
        for step in range(1, g + 1):
            for i in range(length + 1, length * 2 + 1):
                dragonPosition[i - 1] = (dragonPosition[length * 2 - i] + 1) % 4
                
            length *= 2

        MAP[r][c] = 1

        move_count = 1 << g # move = 2^g
        for k in range(move_count):
            r = r + dr[dragonPosition[k]]
            c = c + dc[dragonPosition[k]]
            
            MAP[r][c] = 1

def count_square():
    sum = 0
    for r in range(100):
        for c in range(100):
            if MAP[r][c] + MAP[r + 1][c] + MAP[r][c + 1] + MAP[r + 1][c + 1] == 4:
                sum += 1
    return sum

# T = int(input())
T = 1
for _ in range(T):
    make_dragon_curve()
    
    print(count_square())
